File: arbol_regresion.R
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("ggplot2")
# install.packages("caret")
# install.packages("reshape2")

# Cargar las librerías
library(rpart)        # Para el árbol de regresión
library(rpart.plot)   # Para visualizar el árbol de regresión
library(ggplot2)      # Para visualizaciones
library(caret)        # Para funciones de partición y evaluación
library(reshape2)     # Para transformar datos para visualización

# ---------------------------------------------------
# 2. Cargar y Visualizar el Dataset
# ---------------------------------------------------

# Cargar el dataset
datos <- read.csv('datos_precio_casa.csv')

# Mostrar las primeras filas del dataset
cat("Primeras filas del dataset:\n")
print(head(datos))

# Descripción estadística del dataset
cat("\nDescripción Estadística del Dataset:\n")
print(summary(datos))

# Visualizar la relación entre Edad_Casa y Precio_Casa
ggplot(datos, aes(x = Edad_Casa, y = Precio_Casa)) +
  geom_point(color = 'blue') +
  ggtitle('Relación entre Edad de la Casa y Precio') +
  xlab('Edad de la Casa (años)') +
  ylab('Precio de la Casa (en miles de dólares)') +
  theme_minimal()

# ---------------------------------------------------
# 3. Preparación de los Datos
# ---------------------------------------------------

# Variables independientes y dependiente
X <- datos$Edad_Casa
y <- datos$Precio_Casa

# Dividir el dataset en conjunto de entrenamiento y prueba (80% - 20%)
set.seed(42)  # Para reproducibilidad
trainIndex <- createDataPartition(y, p = 0.8, list = FALSE)
X_train <- X[trainIndex]
y_train <- y[trainIndex]
X_test <- X[-trainIndex]
y_test <- y[-trainIndex]

# Crear data frames para entrenamiento y prueba
train_data <- data.frame(Edad_Casa = X_train, Precio_Casa = y_train)
test_data <- data.frame(Edad_Casa = X_test, Precio_Casa = y_test)

# ---------------------------------------------------
# 4. Entrenamiento del Árbol de Regresión
# ---------------------------------------------------

# Crear el modelo de árbol de regresión con una profundidad máxima de 4
arbol_regresion <- rpart(Precio_Casa ~ Edad_Casa, data = train_data, method = "anova",
                         control = rpart.control(maxdepth = 4))

# Visualizar el árbol de regresión
rpart.plot(arbol_regresion, type = 3, fallen.leaves = TRUE,
           main = "Árbol de Regresión de Precio de Casa")

# Predecir en el conjunto de prueba
y_pred <- predict(arbol_regresion, newdata = test_data)

# ---------------------------------------------------
# 5. Evaluación del Árbol de Regresión
# ---------------------------------------------------

# Calcular métricas de evaluación
mae <- mean(abs(y_test - y_pred))
mse <- mean((y_test - y_pred)^2)
r2 <- 1 - sum((y_test - y_pred)^2) / sum((y_test - mean(y_test))^2)

# Mostrar los resultados
cat("\nEvaluación del Árbol de Regresión:\n")
cat(sprintf("Error Absoluto Medio (MAE): %.2f\n", mae))
cat(sprintf("Error Cuadrático Medio (MSE): %.2f\n", mse))
cat(sprintf("R² Score: %.4f\n", r2))

# Visualizar las predicciones del árbol de regresión
ggplot() +
  geom_point(aes(x = test_data$Edad_Casa, y = test_data$Precio_Casa),
             color = 'blue', alpha = 0.6, label = 'Datos Reales') +
  geom_line(aes(x = test_data$Edad_Casa, y = y_pred),
            color = 'red', size = 1, label = 'Predicciones del Árbol') +
  ggtitle('Árbol de Regresión - Predicciones vs Datos Reales') +
  xlab('Edad de la Casa (años)') +
  ylab('Precio de la Casa (en miles de dólares)') +
  theme_minimal()

# ---------------------------------------------------
# 6. Entrenamiento del Modelo de Regresión Lineal
# ---------------------------------------------------

# Crear y entrenar el modelo de regresión lineal
modelo_lineal <- lm(Precio_Casa ~ Edad_Casa, data = train_data)

# Predecir en el conjunto de prueba con el modelo lineal
y_pred_lineal <- predict(modelo_lineal, newdata = test_data)

# ---------------------------------------------------
# 7. Evaluación del Modelo de Regresión Lineal
# ---------------------------------------------------

# Calcular métricas de evaluación para el modelo lineal
mae_lineal <- mean(abs(y_test - y_pred_lineal))
mse_lineal <- mean((y_test - y_pred_lineal)^2)
r2_lineal <- 1 - sum((y_test - y_pred_lineal)^2) / sum((y_test - mean(y_test))^2)

# Mostrar los resultados
cat("\nEvaluación del Modelo Lineal:\n")
cat(sprintf("Error Absoluto Medio (MAE): %.2f\n", mae_lineal))
cat(sprintf("Error Cuadrático Medio (MSE): %.2f\n", mse_lineal))
cat(sprintf("R² Score: %.4f\n", r2_lineal))

# Visualizar las predicciones del modelo lineal
ggplot() +
  geom_point(aes(x = test_data$Edad_Casa, y = test_data$Precio_Casa),
             color = 'blue', alpha = 0.6, label = 'Datos Reales') +
  geom_smooth(aes(x = test_data$Edad_Casa, y = y_pred_lineal),
              method = 'lm', se = FALSE, color = 'green', size = 1, label = 'Regresión Lineal') +
  ggtitle('Regresión Lineal - Predicciones vs Datos Reales') +
  xlab('Edad de la Casa (años)') +
  ylab('Precio de la Casa (en miles de dólares)') +
  theme_minimal()

# ---------------------------------------------------
# 8. Comparación de Modelos
# ---------------------------------------------------

# Crear un DataFrame para comparar las métricas
comparativa <- data.frame(
  Modelo = c("Árbol de Regresión", "Regresión Lineal"),
  MAE = c(mae, mae_lineal),
  MSE = c(mse, mse_lineal),
  R2_Score = c(r2, r2_lineal)
)

# Mostrar la comparativa
cat("\nComparativa de Métricas entre Modelos:\n")
print(comparativa)

# Visualizar la comparativa de métricas
comparativa_melted <- melt(comparativa, id.vars = "Modelo")

ggplot(comparativa_melted, aes(x = variable, y = value, fill = Modelo)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  ggtitle('Comparación de Métricas entre Modelos') +
  xlab('Métrica') +
  ylab('Valor') +
  theme_minimal() +
  scale_fill_brewer(palette = "Set1")

File: arbol_regresion.py

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Configurar estilo de los gráficos
sns.set(style="whitegrid")

# Cargar el dataset
datos = pd.read_csv('datos_precio_casa.csv')

# Visualizar los datos
plt.figure(figsize=(10, 6))
plt.scatter(datos['Edad_Casa'], datos['Precio_Casa'], color='blue', label='Datos')
plt.title('Relación entre Edad de la Casa y Precio')
plt.xlabel('Edad de la Casa (años)')
plt.ylabel('Precio de la Casa (en miles de dólares)')
plt.legend()
plt.show()

# Variables independientes y dependiente
X = datos[['Edad_Casa']]
y = datos['Precio_Casa']

# Dividir el dataset en conjunto de entrenamiento y prueba
from sklearn.model_selection import train_test_split

X_entrenamiento, X_prueba, y_entrenamiento, y_prueba = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Entrenar el árbol de regresión
from sklearn.tree import DecisionTreeRegressor

# Crear el modelo
arbol_regresion = DecisionTreeRegressor(random_state=42, max_depth=4)

# Entrenar el modelo
arbol_regresion.fit(X_entrenamiento, y_entrenamiento)

# Predecir en el conjunto de prueba
y_pred = arbol_regresion.predict(X_prueba)

# Evaluar el modelo
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mae = mean_absolute_error(y_prueba, y_pred)
mse = mean_squared_error(y_prueba, y_pred)
r2 = r2_score(y_prueba, y_pred)

print("Evaluación del Árbol de Regresión:")
print(f"Error Absoluto Medio (MAE): {mae:.2f}")
print(f"Error Cuadrático Medio (MSE): {mse:.2f}")
print(f"R^2 Score: {r2:.4f}")

# Visualizar las predicciones
# Para visualización, ordenamos los datos
X_prueba_sorted = X_prueba.sort_values(by='Edad_Casa')
y_pred_sorted = arbol_regresion.predict(X_prueba_sorted)

plt.figure(figsize=(10, 6))
plt.scatter(X_prueba['Edad_Casa'], y_prueba, color='blue', label='Datos Reales')
plt.plot(X_prueba_sorted['Edad_Casa'], y_pred_sorted, color='red', label='Predicciones del Árbol')
plt.title('Árbol de Regresión - Predicciones vs Datos Reales')
plt.xlabel('Edad de la Casa (años)')
plt.ylabel('Precio de la Casa (en miles de dólares)')
plt.legend()
plt.show()

# Comparación con Regresión Lineal
from sklearn.linear_model import LinearRegression

# Entrenar el modelo lineal
modelo_lineal = LinearRegression()
modelo_lineal.fit(X_entrenamiento, y_entrenamiento)

# Predecir en el conjunto de prueba
y_pred_lineal = modelo_lineal.predict(X_prueba)

# Evaluar el modelo lineal
mae_lineal = mean_absolute_error(y_prueba, y_pred_lineal)
mse_lineal = mean_squared_error(y_prueba, y_pred_lineal)
r2_lineal = r2_score(y_prueba, y_pred_lineal)

print("\nEvaluación del Modelo Lineal:")
print(f"Error Absoluto Medio (MAE): {mae_lineal:.2f}")
print(f"Error Cuadrático Medio (MSE): {mse_lineal:.2f}")
print(f"R^2 Score: {r2_lineal:.4f}")

# Visualizar las predicciones del modelo lineal
plt.figure(figsize=(10, 6))
plt.scatter(X_prueba['Edad_Casa'], y_prueba, color='blue', label='Datos Reales')
plt.plot(X_prueba['Edad_Casa'], y_pred_lineal, color='green', label='Predicciones Lineales')
plt.title('Regresión Lineal - Predicciones vs Datos Reales')
plt.xlabel('Edad de la Casa (años)')
plt.ylabel('Precio de la Casa (en miles de dólares)')
plt.legend()
plt.show()

# Comparación de modelos
comparativa = pd.DataFrame({
    'Modelo': ['Árbol de Regresión', 'Regresión Lineal'],
    'MAE': [mae, mae_lineal],
    'MSE': [mse, mse_lineal],
    'R^2 Score': [r2, r2_lineal]
})

print("\nComparativa de Métricas entre Modelos:")
print(comparativa)

# Visualizar la comparación
comparativa_melted = comparativa.melt(id_vars='Modelo', var_name='Métrica', value_name='Valor')

plt.figure(figsize=(10, 6))
sns.barplot(x='Métrica', y='Valor', hue='Modelo', data=comparativa_melted)
plt.title('Comparación de Métricas entre Modelos')
plt.ylabel('Valor')
plt.show()
